clang/test/SemaSYCL/union-kernel-param.cpp

```diff
@@ -0,0 +1,47 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -ast-dump %s | FileCheck %s
+// expected-no-diagnostics
+
+// This test checks that compiler generates correct kernel arguments for
+// union without array.
+
+union MyUnion {
+  int x;
+  char y;
+  float cuda;
+};
+
+template <typename name, typename Func>
+__attribute__((sycl_kernel)) void a_kernel(Func kernelFunc) {
+  kernelFunc();
+}
+
+int main() {
+  MyUnion accel;
+
+  a_kernel<class kernel>(
+      [=]() {
+        float local = accel.cuda;
+      });
+}
+
+// Check kernel parameters
+// CHECK: FunctionDecl {{.*}}kernel{{.*}} 'void (MyUnion, int, char, float)'
+// CHECK-NEXT: ParmVarDecl {{.*}} used _arg_ 'MyUnion'
+// CHECK-NEXT: ParmVarDecl {{.*}} used _arg_x 'int'
+// CHECK-NEXT: ParmVarDecl {{.*}} used _arg_y 'char'
+// CHECK-NEXT: ParmVarDecl {{.*}} used _arg_cuda 'float'
```

Comment:
Looks like decomposition is happening here. arg_x, arg_y, and arg_cuda shouldn't end up as kernel parameters. 
For some reason, handlers are not visiting correctly.

@Fznamznon and @elizabethandrews could you please look at the latest patches here? Any suggestion would be greatly appreciated. Thanks.

Reply:
See https://github.com/intel/llvm/pull/2285#discussion_r469786490 .

Reply:
Done